### ISSUE TITLE
Support for `uint16`, `uint32`, and `uint64`

### ISSUE THREAD

**Issue by pmeier** [pytorch/pytorch#58734]
The [array API specification](https://data-apis.org/array-api/latest/API_specification/data_types.html) stipulates the data types that we need to support to be compliant. Currently we are missing support for `uint16`, `uint32`, and `uint64`.

**Comment by rgommers** [comment]
Someone just asked about other uint types on Slack. And from https://github.com/pytorch/vision/pull/4326#discussion_r697371491:

> This is because PyTorch doesn't (yet) support `uint16`, and is also a problem when reading PNG images of type `uint16`.

16-bit image support making `uint16` the most interesting one of the missing dtypes had been my guess before.

There are no plans to work on this issue currently I think, unless more demand materializes.

**Comment by NicolasHug** [comment]
Another tangible need for uint16 support in torchvision is https://github.com/pytorch/vision/issues/4731

We added support for native 16 bits png decoding in torchvision, but we can't make this API public for now, because we output int32 tensors and this wouldn't be compatible with the rest of our transforms.
It'd be great if we could make it public because Pillow 16 bits png support is fairly limited.

**Comment by kernelmethod** [comment]
Bumping this.

My research collaborators and I are working on some cryptographic applications where we could really use `uint32` / `uint64`. Some operations on Z_{2^64}, e.g. comparison, would be a lot more straightforward to implement with unsigned integers as the underlying dtype.

**Comment by a-gn** [comment]
We have a GIS pipeline that uses image transforms from multiple projects, some of which only support uint, but uint8 leads to too much loss of color information. We could really use uint16 support.

**Comment by neelnanda-io** [comment]
I would appreciate uint16 support! I'm trying to do NLP stuff with a large dataset of tokens between 0 and 51000, and it's annoying to consume double the storage to keep them as int32s (I'm currently storing them as uint16 via HuggingFace, but I need to load them as NumPy and manually convert them)

**Comment by oliver-batchelor** [comment]
I'm doing work on HDR imaging and we read images from the camera as 16-bit unsigned. It's possible to work around it by using other frameworks but it would be really useful.

**Comment by VladShtompel** [comment]
> I'm doing work on HDR imaging and we read images from the camera as 16-bit unsigned. It's possible to work around it by using other frameworks but it would be really useful.

This is exactly the issue me and my team are faced with right now.

**Comment by StrongChris** [comment]
I'm doing work with DICOM data that is often 10 or even 14 unsigned bits. A uint16 would be very nice for these! My work is focused on speed so using the smallest possible datatype would be very appreciated.

**Comment by ezyang** [comment]
We should add these dtypes, and then build out support via PT2. We probably aren't going to add kernels for everything but Triton makes it very easy to JIT compile these operations.

**Comment by NicolasHug** [comment]
@ezyang would Triton be able to enable CPU support?

**Comment by ezyang** [comment]
Not Triton per se, but we have a CPU inductor backend, so the answer is yes!

**Comment by soulitzer** [comment]
From Triage review: We still need some limited eager support, e.g. factory functions, conversion functions. Also consideration with autocast? (maybe not too bad?)

**Comment by vadimkantorov** [comment]
Also, bit ops are only well-defined/standardized in CPUs for unsigned dtypes if I understand well: https://github.com/pytorch/pytorch/issues/105465

**Comment by vadimkantorov** [comment]
uint16 would also be useful for interop with opencv (`CV_16U` dtype)

**Comment by tchaton** [comment]
Hey, torch.uint16 would be good to encode text into tokens to reduce memory footprint from  uint32 when the vocab isn't too big.

**Comment by smdrnks** [comment]
+1, also have a language modelling use case where uint16 could save quite some memory. Would be great to have this.

**Comment by DrDryg** [comment]
I would also appreciate support for uint. Im developing software using for image processing libtorch as a backend and it would be very useful with support for uint. In particular uint16 but uint32 and uint64 would be nice too.

**Comment by vadimkantorov** [comment]
A related issue on having uint16 images:
- https://github.com/pytorch/vision/issues/8188

**Comment by ezyang** [comment]
Some dumb problems we will have to work out.

* `uint8` tensor accumulates into `int64` tensor, which makes sense when you don't have int64 tensor but makes a lot less sense when you have uint64 tensor. This leads to a potential inconsistency with the larger types; in particular, it is an incredibly bad idea for `uint64` to accumulate into `int64`, and `uint32` is probably not a good idea either. A short term stop gap is to leave sum unimplemented but chances are someone will come asking for it. By the way, this is a use case for defining arithmetic operations on our bits types: uint should have a wider accumulate type to prevent overflow, but bits would never widen and always do modular arithmetic.

**Comment by vadimkantorov** [comment]
I think it's better to go ahead and have some dtype representations even if meaningful ops are not supported at first and only conversions/casts/reinterprets/restride are implemented: mainly for expected interop and faithfulness of representation. As long as there is a dedicated docs page for that dtype with explained quirks, I think it's fine

Same reasoning might apply for the following :)
- https://github.com/pytorch/pytorch/issues/104702

**Comment by vadimkantorov** [comment]
Regarding sum, maybe a transitory option might be to require explicit args specifying the out_dtype and acc_dtype? (it would also be nice to elide temporary full upcasting allocations 
 https://github.com/pytorch/pytorch/issues/55366)

**Comment by ionutmodo** [comment]
Hi guys! I would like to add another usage for uint16, which can be used in designing efficient adaptive sparse optimizers.

**Comment by rbelew** [comment]
ANother `uint64` use case arises trying to embed hashes from  [spaCy](https://spacy.io/)'s token hashing function. Spacy uses hashing on texts to get unique ids (cf. [SO](https://stackoverflow.com/a/53213295/1079688) )

```
>>> import spacy
>>> nlp = spacy.load('en')
>>> text = "here is some test text"
>>> doc = nlp(text)
>>> [token.norm for token in doc]
[411390626470654571, 3411606890003347522, 7000492816108906599, 1618900948208871284, 15099781594404091470]
```

**Comment by vadimkantorov** [comment]
Another uint16 usecase is for representing UTF-16 code points... e.g. as one of ways of interop/representing arrays of stings as tensors:
- https://github.com/pytorch/pytorch/issues/101699

**Comment by vadimkantorov** [comment]
And another usecase is bitshifts, as bitshift for signed dtypes (if signed bit is set) is theoretically implementation-defined or undefined

**Comment by oliver-batchelor** [comment]
@vadimkantorov @rbelew In case you didn't realise torch >= 2.3 has all the unsigned int types.

**Comment by vadimkantorov** [comment]
Yeah, I saw the support, but I didn't realize which ops are enabled for them. E.g. is bitshift enabled?

**Comment by ezyang** [comment]
@malfet just added bitshift. Op support is on a per request basis: we will add it if requested

**Comment by vadimkantorov** [comment]
> just added bitshift.

If you are meaning the https://github.com/pytorch/pytorch/pull/135525, it only added and/or/xor, but maybe bitshifts were added in some other PRs...

**Comment by FL33TW00D** [comment]
Echoing the comment from @neelnanda-io, would be great to add `uint16` support for `Embedding` indices. A good chunk of vocabs are smaller than 65535 :)

**Comment by ezyang** [comment]
for embedding I think this involves beefing up index_select

**Comment by vadimkantorov** [comment]
related on supporting natively shorted bitwidth indexes:
- https://github.com/pytorch/pytorch/issues/61819

**Comment by vadimkantorov** [comment]
> @malfet just added bitshift.

@ezyang it appears that not yet:
- https://github.com/pytorch/pytorch/issues/139124

**Comment by malfet** [comment]
> > @malfet just added bitshift.
> 
> @ezyang it appears that not yet:
> 
> * [shift operators not supporting integral types #139124](https://github.com/pytorch/pytorch/issues/139124)

Yeah, It's not part of 2.5.0 release branch, but should be present in nightly...

**Comment by ev-br** [comment]
As a data point, we tried enabling unsigned integers in the Array API test suite, in https://github.com/data-apis/array-api-compat/pull/253

That results in 70-odd failures, the majority of which are of the `*_cpu is not implemented for uint16` variety.

**Comment by mdhaber** [comment]
I also ran accross those `*_cpu is not implemented for uint*` errors (https://github.com/pytorch/pytorch/issues/58734#issuecomment-2684450868) when testing `marray` (array API masked arrays) with PyTorch as the backend (https://github.com/mdhaber/marray/pull/110). 

I thought I'd also report:
```python3
import torch as xp
x = xp.asarray([1, 2, 3], dtype=xp.uint64)
sentinel = xp.iinfo(xp.uint64).max // 2 + 1
x[0] = sentinel - 1  # OK
x[0] = xp.asarray(sentinel, dtype=x.dtype)  # OK
x[0] = sentinel  # RuntimeError: Overflow when unpacking long
```

**Comment by egeboy35** [comment]
Requesting `cumsum`/`cumprod` (and asking a design question about `sum`/`prod`)
for the barebones unsigned types, per the "op support is on a per request basis"
note above.

**What is actually broken.** The default path already works: `torch.cumsum(x)`
on a uint16/32/64 tensor promotes to int64 via `get_dtype_from_self`
(`ReduceOpsUtils.h`), so it returns an int64 result today. What fails is the
explicit path — `dtype=torch.uint32` or `out=<uint32 tensor>` raises
`"cumsum_out_cpu" not implemented for 'UInt32'`, because the kernels still
dispatch through legacy macros:

* `aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:84` (cumsum), `:103` (cumprod)
* `aten/src/ATen/native/cuda/CumsumKernel.cu:11`, `CumprodKernel.cu:11`

**Proposed change (cumsum/cumprod only).** Convert those four sites to
`AT_DISPATCH_V2` and append `AT_EXPAND(AT_BAREBONES_UNSIGNED_TYPES)`, exactly
the shape of #189953 (clamp/min/max, landed 2026-08-11). One extra piece is
needed: the CPU kernels accumulate via `at::acc_type<scalar_t, false>`, which
has no row for these dtypes, so `AccumulateType.h` needs
`uint16_t/uint32_t/uint64_t -> uint64_t`. I picked unsigned accumulation
deliberately — `int64_t` (what the signed integral rows use) cannot represent
the upper half of the uint64 range.

**Design question for `sum`/`prod`, which I deliberately left out.** Those hit
the accumulator question raised in this thread (uint64 accumulating into int64
being "an incredibly bad idea"). Widening `SumKernel.cpp:616` /
`ReduceOpsKernel.cpp:164` / `cuda/ReduceSumProdKernel.cu:229,257` is mechanically
the same edit, but it needs a decision on the accumulator type first, and it also
trips `test_result_dtype` in `test/test_reductions.py`, whose `is_integral` check
does not know about barebones unsigned. Happy to do it as a follow-up once
there's a call on the accumulator.

Is #189953 enough precedent for me to send the cumsum/cumprod PR, or should I
wait for an `actionable` label here? The change is written and the CPU kernels
compile clean locally in all three ISA variants; I can't exercise the runtime
path on this machine, so the behavioural test would be validated in CI.
@Skylion007 you merged #189953 — would you be the right reviewer?

**Comment by ngimel** [comment]
No, #189953 is not enough precedent for this. #189953 explicitly states "With clamp supporting these dtypes, those tests pass and the xfails can be removed there." as motivation. Removing xfails is not a user request! Barebones types are barebones for a reason - we don't want to support op creep on them.

**Comment by egeboy35** [comment]
Understood, and thanks for the quick, clear answer — dropping it.

That distinction is the useful part: I read #189953 as a general dtype-coverage
precedent, when its motivation was specific to unblocking those xfails. Noted
that the barebones types are deliberately kept narrow rather than widened on
request.
